### PR TITLE
Convert BaseJavaModuleTest to Kotlin

### PR DESCRIPTION
**Pull request by abdennour-jebbar-nw**
## Summary:
As part of the effort to Kotlin-fy React Native tests, I've converted [BaseJavaModuleTest](https://github.com/facebook/react-native/blob/main/packages/react-native/ReactAndroid/src/test/java/com/facebook/react/bridge/BaseJavaModuleTest.java) to Kotlin.

## Changelog:
[Internal] [Changed] - Convert BaseJavaModuleTest to Kotlin

## Test Plan:
Tests pass: ./gradlew :packages:react-native:ReactAndroid:test
Formatted with [KtFmt](https://facebook.github.io/ktfmt/)

### CODE REVIEW

File: packages/react-native/ReactAndroid/src/test/java/com/facebook/react/bridge/BaseJavaModuleTest.kt [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    var posn = -1
+    for (i in methods!!.indices) {
+      val md = methods[i]
+      if (md.name === mname) {
+        posn = i
+        break
+      }
+    }
+    return posn
+  }
```
Comment (rshest) [inline packages/react-native/ReactAndroid/src/test/java/com/facebook/react/bridge/BaseJavaModuleTest.kt:50]:
```suggestion
  private fun findMethod(mname: String, methods: List<JavaModuleWrapper.MethodDescriptor>): Int = methods.indexOfFirst({it.name === mname})
```

File: packages/react-native/ReactAndroid/src/test/java/com/facebook/react/bridge/BaseJavaModuleTest.kt [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+  @Test(expected = NativeArgumentsParseException::class)
+  @Throws(Exception::class)
```
Comment (rshest) [inline packages/react-native/ReactAndroid/src/test/java/com/facebook/react/bridge/BaseJavaModuleTest.kt:53]: This should be removed.

File: packages/react-native/ReactAndroid/src/test/java/com/facebook/react/bridge/BaseJavaModuleTest.kt [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+  }
+
+  private abstract inner class NativeTestGeneratedModuleSpec : BaseJavaModule(), TurboModule {
+    @ReactMethod abstract fun generatedMethod(a: String?, b: Int)
+  }
+
+  private inner class GeneratedMethodsModule : NativeTestGeneratedModuleSpec() {
+    override fun getName(): String {
+      return "GeneratedMethods"
+    }
```
Comment (rshest) [inline packages/react-native/ReactAndroid/src/test/java/com/facebook/react/bridge/BaseJavaModuleTest.kt:111]:
```suggestion
    override fun getName(): String = "GeneratedMethods"
```

**Comment by rshest** [comment]
@abdennour-jebbar-nw 

Thank you so much for taking this on!

Looks good overall, aside from a few stylistic suggestions.

**Review by rshest — CHANGES_REQUESTED** [review]
Could you please address the stylistic suggestions? Thanks!

File: packages/react-native/ReactAndroid/src/test/java/com/facebook/react/bridge/BaseJavaModuleTest.kt (hunk 2 above)
Comment (abdennour-jebbar-nw) [reply packages/react-native/ReactAndroid/src/test/java/com/facebook/react/bridge/BaseJavaModuleTest.kt:43]: Thank you for your reviews, I would love to know why this should be removed

File: packages/react-native/ReactAndroid/src/test/java/com/facebook/react/bridge/BaseJavaModuleTest.kt [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    moduleWrapper = JavaModuleWrapper(null, moduleHolder)
+    methods = moduleWrapper.methodDescriptors
+    val generatedModuleHolder = ModuleHolder(GeneratedMethodsModule())
+    generatedModuleWrapper = JavaModuleWrapper(null, generatedModuleHolder)
+    generatedMethods = generatedModuleWrapper.methodDescriptors
+    arguments = mock(ReadableNativeArray::class.java)
+  }
+
+  private fun findMethod(mname: String, methods: List<JavaModuleWrapper.MethodDescriptor>?): Int = methods.indexOfFirst({it.name === mname})
+
```
Comment (rshest) [inline packages/react-native/ReactAndroid/src/test/java/com/facebook/react/bridge/BaseJavaModuleTest.kt:41]: This should be `List<JavaModuleWrapper.MethodDescriptor>` (without `?` at the end).

File: packages/react-native/ReactAndroid/src/test/java/com/facebook/react/bridge/BaseJavaModuleTest.kt (hunk 2 above)
Comment (rshest) [reply packages/react-native/ReactAndroid/src/test/java/com/facebook/react/bridge/BaseJavaModuleTest.kt:43]: The `@Test(expected = NativeArgumentsParseException::class)` bit should be enough to fulfill the purpose of this test, the `@Throws(Exception::class)` part doesn't add anything there in terms of information, neither it's required for proper compilation.

**Review by rshest — CHANGES_REQUESTED** [review]
@abdennour-jebbar-nw 

Sending it back to you, once again (sorry!), since `List<JavaModuleWrapper.MethodDescriptor>?` breaks the build - should be `List<JavaModuleWrapper.MethodDescriptor>`.

**Review by rshest — APPROVED** [review]

**Comment by abdennour-jebbar-nw** [comment]
@cortinico I'm investigating the failing tests

**Comment by rshest** [comment]
Hey @abdennour-jebbar-nw , what are your plans with the diff? 

Will you be able to follow up, or should we maybe close it?

**Comment by abdennour-jebbar-nw** [comment]
@rshest I'm still following up with this PR, I'm trying to figure out why the moduleWrapper remains null

File: packages/react-native/ReactAndroid/src/test/java/com/facebook/react/bridge/BaseJavaModuleTest.kt [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+  }
+
+  private abstract inner class NativeTestGeneratedModuleSpec : BaseJavaModule(), TurboModule {
+    @ReactMethod abstract fun generatedMethod(a: String, b: Int)
+  }
+
+  private inner class GeneratedMethodsModule : NativeTestGeneratedModuleSpec() {
+    override fun getName(): String  ="GeneratedMethods"
+
+    override fun generatedMethod(a: String, b: Int) {}
```
Comment (rshest) [inline packages/react-native/ReactAndroid/src/test/java/com/facebook/react/bridge/BaseJavaModuleTest.kt:96]:
```suggestion
    override fun generatedMethod(a: String?, b: Int?) {}
```

File: packages/react-native/ReactAndroid/src/test/java/com/facebook/react/bridge/BaseJavaModuleTest.kt [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    @ReactMethod fun regularMethod(a: String, b: Int) {}
+
+    @ReactMethod fun asyncMethod(a: Int, p: Promise) {}
+
+    @ReactMethod(isBlockingSynchronousMethod = true)
+    fun syncMethod(a: Int, b: Int): Int = a + b
+  }
+
+  private abstract inner class NativeTestGeneratedModuleSpec : BaseJavaModule(), TurboModule {
+    @ReactMethod abstract fun generatedMethod(a: String, b: Int)
```
Comment (rshest) [inline packages/react-native/ReactAndroid/src/test/java/com/facebook/react/bridge/BaseJavaModuleTest.kt:90]:
```suggestion
    @ReactMethod abstract fun generatedMethod(a: String?, b: Int?)
```

File: packages/react-native/ReactAndroid/src/test/java/com/facebook/react/bridge/BaseJavaModuleTest.kt [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+  fun testCallGeneratedMethod() {
+    val methodId = findMethod("generatedMethod", generatedMethods)
+    whenever(arguments.size()).thenReturn(2)
+    generatedModuleWrapper.invoke(methodId, arguments)
+  }
+
+  private class MethodsModule : BaseJavaModule() {
+    override fun getName(): String = "Methods"
+
+    @ReactMethod fun regularMethod(a: String, b: Int) {}
```
Comment (rshest) [inline packages/react-native/ReactAndroid/src/test/java/com/facebook/react/bridge/BaseJavaModuleTest.kt:81]:
```suggestion
    @ReactMethod fun regularMethod(a: String?, b: Int?) {}
```

**Comment by abdennour-jebbar-nw** [comment]
@rshest Thank you so much for following up with me, I'll try the changes asap, once green I'll update the PR. Your help was so much appreciated 🙌

**Comment by rshest** [comment]
> @rshest Thank you so much for following up with me, I'll try the changes asap, once green I'll update the PR. Your help was so much appreciated 🙌

@abdennour-jebbar-nw I verified internally that the tests are fixed, so your PR is merged to trunk now. Thank you very much, once again, for working on this! :)

**Pull request closed without merging**